File: 078__Subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def iter(nums, result):
            if len(nums) == 0:
                return result
            result += [x + [nums[0]] for x in result]
            return iter(nums[1:], result)

        return iter(nums, [[]])
            

File: Java/078__Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        result.add(new ArrayList<>());
        for (int val : nums) {
            int size = result.size();
            for (int i = 0; i < size; i++) {
                List<Integer> list = new ArrayList<>(result.get(i));
                list.add(val);
                result.add(list);
            }
        }
        return result;
    }
}

// Backtracing Solution
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtracking(result, new ArrayList<>(), 0, nums);
        return result;
    }

    private void backtracking(List<List<Integer>> result, List<Integer> temp_list, int index, int[] nums) {
        result.add(new ArrayList<>(temp_list));
        for (int i = index; i < nums.length; i++) {
            temp_list.add(nums[i]);
            backtracking(result, temp_list, i + 1, nums);
            temp_list.remove(temp_list.size() - 1);
        }
    }
}
